strawberry/channels/handlers/base.py

```diff
@@ -0,0 +1,149 @@
+import asyncio
+import contextlib
+from collections import defaultdict
+from typing import Any, Awaitable, Callable, DefaultDict, Dict, List, Optional, Sequence
+
+from typing_extensions import Literal, Protocol, TypedDict
+
+from channels.consumer import AsyncConsumer
+from channels.generic.websocket import AsyncJsonWebsocketConsumer
+from strawberry.channels.context import StrawberryChannelsContext
+
+
+class ChannelsMessage(TypedDict, total=False):
+    type: str
+
+
+class ChannelsLayer(Protocol):  # pragma: no cover
```

Comment:
It is not possible because channels expects an object that implements the layers interface. Most commonly it will be [redis layer](https://github.com/django/channels_redis/blob/main/channels_redis/core.py#L72), but one can define his own layer class.

I only defined this protocol for typing, based on the [layer spec](https://channels.readthedocs.io/en/stable/channel_layer_spec.html), since [BaseChannelLayer](https://github.com/django/channels/blob/main/channels/layers.py#L95) doesn't implement those and thus would lead to typing issues if typing `channel_layer: Optional[BaseChannelLayer]` below at line 59.

Reply:
but you would need to implement all these methods when using it with Strawberry, no? :)

Reply:
@patrick91 actually no =P. I use the layer based on the spec for `channel_listen`, so I wrote the protocol to have a proper typing for it and avoid `# type:ignore` comments.

But the decision for the layer implementation is up to the user. He can use redis or even an in memory implementation which is the default one provided by channels itself.

What we are implementing is a custom consumer, which its base class sets up a layer object to use when one is installed (it is also possible to not have it installed. I added an error in this case inside `channel_listen`, based on our chat on discord =P)

Reply:
In summary, this protocol is here just to fix poor typing on channels side =P